### PR TITLE
library: Add Grid entry

### PR DESCRIPTION
**Pull request by bazylevnik0**

### CODE REVIEW

**Review by sonnyp — CHANGES_REQUESTED** [review]
Good start!

We try to make Library entries nice to look at and interactive (but still comprehensible).

What about a (very) simple tic tac toe with a 3x3 grid? Since it will require some JavaScript it's also a good opportunity to showcase the APIs to work with grid.

Don't worry too much about style, we can iterate later.

File: src/about.js [hunk 1]
```diff
@@ -56,6 +56,7 @@ ${getBlueprintVersion()}
     "Hari Rana (TheEvilSkeleton) https://theevilskeleton.gitlab.io",
     "Sriyansh Shivam https://linktr.ee/sonic_here",
     "Angelo Verlain https://www.vixalien.com",
+    "bazylevnik0 https://github.com/AkshayWarrier",
```
[the remark this answers is not in the document]
Comment (bazylevnik0) [reply src/about.js:59]: Yep :)

File: src/Library/demos/Grid/main.blp [hunk 2]
```diff
@@ -0,0 +1,71 @@
+using Gtk 4.0;
+using Adw 1;
+
+Adw.StatusPage {
+  title: "Grid";
+  description: _("Arrange widgets in rows and columns.\n\n*Tic Tac Toe: just click to the '#' for 'x' and click again for 'o':");
```
Comment (sonnyp) [inline src/Library/demos/Grid/main.blp:6]:
```suggestion
  description: _("Arrange widgets in rows and columns.");
```

hopefully we won't need to explain the UI :)

**Review by sonnyp — COMMENTED** [review]
I should have been more explicit. With tic tac toe - I meant as a (very simple) game you can play against the computer.

Similar to https://playtictactoe.org/

For the UI please make each cell square.
You can use `circle-outline-thick-symbolic` and `cross-large-symbolic` icons

If you make it look like a tic tac toe - we don't need to explain :)

File: src/Library/demos/Grid/main.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "Grid",
+  "category": "user_interface",
+  "description": "Allow people to perform using a grids",
+  "panels": [
+    "code",
+    "style",
+    "ui",
+    "preview"
```
Comment (sonnyp) [inline src/Library/demos/Grid/main.json:9]: Should be enough to see quickly what Grid is about:

```suggestion
    "ui",
    "preview"
```

**Review by sonnyp — COMMENTED** [review]
Cool.

Regarding the design:

1. No need for edit mode
2. Have one of the cell by default O or X so that it's clear for the user what they're looking at
3. Make the cells square
4. Only show big  O / X, no background

The icons should be visible and light and dark modes - if you stick to the default color that should work fine

File: src/Library/demos/Grid/main.css [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+#button01:hover,
+#button02:hover,
+#button10:hover,
+#button11:hover,
+#button12:hover,
+#button20:hover,
+#button21:hover,
+#button22:hover {
+  box-shadow: inset 0px 0px 65px #ff1493;
+}
```
Comment (sonnyp) [inline src/Library/demos/Grid/main.css:26]: the default `hover` is fine, you can remove this

```suggestion
```

File: src/Library/demos/Grid/main.css [hunk 5]
```diff
@@ -0,0 +1,27 @@
+#button00,
+#button01,
+#button02,
+#button10,
+#button11,
+#button12,
+#button20,
+#button21,
+#button22 {
```
Comment (sonnyp) [inline src/Library/demos/Grid/main.css:9]: in CSS, you should use a class to to make styling reusable between different elements

```suggestion
.cell {
```

You can apply the class to each button using the `styles ['cell']` syntax

And remove `name` on all `button`

File: src/Library/demos/Grid/main.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "Grid",
+  "category": "user_interface",
+  "description": "Allow people to perform using a grids",
```
Comment (sonnyp) [inline src/Library/demos/Grid/main.json:4]:
```suggestion
  "description": "Arrange widgets in rows and columns.",
```

File: src/Library/demos/Grid/main.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+let step = 1;
+function onClicked(button) {
+  //check access for user action
+  let clicked = button.get_child();
+  if (
+    clicked.icon_name !== "cross-large-symbolic" &&
+    clicked.icon_name !== "circle-outline-thick-symbolic"
+  ) {
```
Comment (sonnyp) [inline src/Library/demos/Grid/main.js:25]: an early return would make the function easier to read and reduce one level of indentation

something like

```suggestion
  if (clicked.icon_name) return;
```

File: src/Library/demos/Grid/main.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+for (const id of button_ids) {
+  const button = workbench.builder.get_object(id);
+  button.connect("clicked", onClicked);
+}
+
+let step = 1;
+function onClicked(button) {
+  //check access for user action
+  let clicked = button.get_child();
```
Comment (sonnyp) [inline src/Library/demos/Grid/main.js:21]: let's name it after what it is for clarity

```suggestion
  let image = button.get_child();
```

File: src/Library/demos/Grid/main.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  if (
+    clicked.icon_name !== "cross-large-symbolic" &&
+    clicked.icon_name !== "circle-outline-thick-symbolic"
+  ) {
+    //store and show user action
+    clicked.icon_name = "cross-large-symbolic";
+    //calculate pc reaction
+    let pc_is_thinking = true;
+    let pc_is_thinking_row;
+    let pc_is_thinking_col;
```
Comment (sonnyp) [inline src/Library/demos/Grid/main.js:31]: these don't need to be scope outside the `while`

File: src/Library/demos/Grid/main.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+      pc_is_thinking_col = "" + Math.floor(Math.random() * 3);
+      //make pc reaction if possible
+      let temp = workbench.builder.get_object(
+        "button" + pc_is_thinking_row + pc_is_thinking_col,
+      );
+      let searched = temp.get_child();
+      if (
+        searched.icon_name !== "cross-large-symbolic" &&
+        searched.icon_name !== "circle-outline-thick-symbolic"
+      ) {
```
Comment (sonnyp) [inline src/Library/demos/Grid/main.js:43]: ditto

File: src/Library/demos/Grid/main.blp [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+  description: _("Arrange widgets in rows and columns.");
+
+ Box {
+    halign: center;
+    Grid {
+        Button button00 {
+        name: "button00";
+        layout { row: 0; column: 0; }  
+        hexpand: false;
+        Image image00{
```
Comment (sonnyp) [inline src/Library/demos/Grid/main.blp:15]:
```suggestion
        Image {
```

you can remove all ids on image, they're not used

**Review by sonnyp — CHANGES_REQUESTED** [review]
Looks great now!

Just a few more changes and this is ready to go.

Try to make sure the code is always easy to reason about.

**Review by sonnyp — APPROVED** [review]
The point if early return is that you don't need the `else` :)

I fixed it in https://github.com/sonnyp/Workbench/pull/233/commits/810fadfe75bfd6fc5de85e4c7386b465d7c73d49

Thanks for the contribution!

**Pull request merged**